wordpress-mobile/AztecEditor-Android
# Issue/555 buffered action detection

Fixes #530, #555

This PR implements the buffered action mechanism where events are observed and collected to detect known patterns for actions that produce more than 1 TextWatcher event per each individual user action, as otherwise implicitly expected by Aztec.

### Essentials
Instead of having all TextWatchers added directly to AztecText, we now have only one Android TextWatcher (which is implemented by AztecText itself) and we keep an internal list of other TextWatchers ourselves (`observedWatchers`). 

The original idea was to collect data from the received TextWatcher events (in their before, on, and after moments) and finally decide whether these events correspond to a known pattern such as the one described in #530 (particularly, inserting a space in a word on API26 and later is composed of 4 events: 2 deletes, and 2 inserts, instead of just 1 event that shows the space addition as can be seen on platforms other than API26+).
Then, if the pattern was detected, make up a new event that would do the same thing as in other platforms, so to standardise behaviour across Android versions, and instead send this made-up event to the regular TextWatchers that are awaiting for new events to come and be processed. All other non-corresponding events would just get processed as usual.

Unfortunately I discovered that this was impossible, as the Android’s `TextView` `TextWatcher` interface does not prevent things from being executed in any way, it just tells the listeners that something is about to happen (`beforeTextChanged`), happening (`onTextChanged`), and has just happened (`afterTextChanged`).

I decided to tweak the initial idea a little bit and, instead of trying to "_make the TextWatchers wait_",  we now only observe events happen (and dispatch them right away), and once the pattern is detected we quickly move on to operate on the content and produce the desired output.

###  Key components
- A `TextWatcherEvent` serves as a placeholder where to keep data that comes from Android’s TextWatcher interface (before, on, and after data).
- A `UserOperationEvent` is pretty much a sequence of `TextWatcherEvent`s
- A `Bucket` may contain several defined `UserOperation`s
- A specific bucket may belong to a specific platform, a range of platforms, or even a specific device model. (we now only have the implementation for API26 as described in #530 as it's the only detected case for now)
- The `ObservationQueue` is a sequence, and is filled with `TextWatcherEvents` as they come, and “piled up” if each event corresponds to an expected pattern in a known bucket, or if some divergence is found (that is, the incoming events don’t seem to match any of the expected patterns) then the queue is immediately cleared.
It is important to know that this queue is filled in with events that are very close together (under 50 milliseconds between each other), to avoid queuing up elements that are genuine user operations.
The way we do this is by comparing the time in which the events are created and check the difference between one another. We don’t use Timers nor any other elaborate structure to do that, to avoid introducing any possibility for indeterminate things to happen.
The check is also so simple, that there is no way for the queue to pile up more events than what any of the known sequences’s length in the selected buckets. For the matter of this PR which one has 1 bucket and 1 sequence implemented, this means it will only pile up up to 4 events (if they partially match the expected sequence as they come), given the pattern we observed in #530 is composed of exactly 4 events. This way, as we make sure to be looking into ALL and EVERY event, we don’t need to look into partial sequences or try to figure out or make sense out of the buffer content.

It is also important to note that the Bucket system makes the checks only available to whatever platform it was made for, so for instance running this on API17 will never enqueue anything, so no possibility to introduce new bugs on other platforms (which is something we have been running risks for before, as we would always try to find a generic way for all platforms).



### How it works

AztecText is the only TextWatcher the Android platform knows for AztecText view itself.

All of the events data is collected by it and then built into a `TextWatcherEvent` by using the builder pattern in the `afterTextChanged` method override, as we know this is the last of the 3 moments in the interface. Note that all the _before_, _on_, and _after_ events are passed to the whatever `TextWatchers` are registered to AztecText as well, as we can’t really prevent things from being executed (we can only “see” what’s going on). 

Once a `TextWatcherEvent` is built, it is immediately added to the `ObservationQueue`, which processes the event upon addition to check if a pattern corresponding to defined `UserOperation`s is building up.

The `ObservationQueue` then checks if we know this pattern or not. As soon as a queue diverges from all known patterns, it is cleared up and the queue starts to build up again with the new TextWatcher event. If it matches, it’s kept until a full pattern is found. Once found, the `UserOperationEvent` `buildReplacementEventWithSequenceData()` method is called, and with this data we end up calling the `executeEvent()` method of the registered `IEventInjector`.

The implementation of `IEventInjector` is in charge of actually operating on the content to make the desired operation happen in the way that it was expected to work.



### Other notes

- The AztecText TextWatcher implementation was caring for history keeping, we moved that into an anonymous watcher that is added in method `addHistoryLoggingWatcher`.
- Some of the code is Kotlin, some of it is still Java as I didn’t have enough expertise to cope with the complexity of this solution and deal with learning Kotlin at the same time (I tried, but was worried about time spent). I expect to move whatever is Java to Kotlin on another PR if possible. We could even wait to merge this until another Java-to-Kotlin PR gets merged against this branch as well.



### To test:
On API26+
1. on an empty AztecText instance, type “he”
2. tap on Bold
3. type “llo”
4. place the cursor right before “llo”
5. tap the space key
6. observe the format for each part of the word is kept as intended (the “he” part remains unstyled, while the “llo” part remains in Bold style).

Repeat the same on other platforms and observe the same behaviour.


cc @0nko , @daniloercoli , @maxme
